tests/s3_bucket/test_s3_bucket_books.py

```diff
@@ -0,0 +1,136 @@
+    code_tag = code_tag
+
+    # queue filename set in the bakery environment, same for both staging and production
+    queue_filename = "distribution-queue.json"
+
+    json_output_filename = f"{os.getcwd()}/fixtures/data/webview/s3_bucket_books.json"
+
+    s3_bucket_books = []
+
+    queue_state_key = f"{code_tag}.{queue_filename}"
+
+    resp = client.list_object_versions(Bucket=s3_queue_state_bucket, Prefix=queue_state_key)
+
+    versions = resp["Versions"]
+
+    for version in versions:
+        resp = client.get_object(
+            Bucket=s3_queue_state_bucket, Key=queue_state_key, VersionId=version["VersionId"]
+        )
+        s3_bucket_books.append(json.loads(resp["Body"].read()))
+
+    # writes the json file with required book details (see the top message)
+    with open(json_output_filename, "w") as json_output_file:
+        json.dump(s3_bucket_books, json_output_file)
+
+
+def test_s3_bucket_books(s3_base_url, code_tag, s3_queue_state_bucket_books):
+
+    # path to the aws s3 bucket folder
+    s3_archive_folder = f"/apps/archive/{code_tag}/contents/"
+
+    # reading nested lists and getting slug names
+    json_data = json.loads(s3_queue_state_bucket_books)
+    json_data = json_data[::-1]
+
+    uuid_list = []
+    version_list = []
+    for key in json_data:
+        if key["uuid"]:
+            uuid_list.append(key["uuid"])
+        if key["version"]:
+            version_list.append(key["version"])
+
+    for i in range(0, len(version_list)):
+        version_list[i] = version_list[i][2:]
+
+    approved_books_full_url = [
+        f"{s3_base_url}{s3_archive_folder}{i}@{j}.json" for i, j in zip(uuid_list, version_list)
+    ]
+
+    successful_books = []
+    unsuccessful_books = []
+
+    for url in approved_books_full_url:
+        try:
+            urllib.request.urlopen(url)
+        except HTTPError as h_e:
+            # Return code 404, 501, ...
+            print("HTTPError (check concourse jobs): {}".format(h_e.code) + ", " + url)
+            unsuccessful_books.append(url)
+
+        else:
+            successful_books.append(url.replace(".json", ".xhtml"))
+
+    # list of approved book urls without http errors (unsuccessful concourse jobs)
+    for book in successful_books:
+
+        print("Verifying pages of collection ", book)
+
+        xhtml_data = requests.get(book).content
+        doc = etree.fromstring(xhtml_data)
+
+        links = []
+        for node in doc.xpath(
+            '//x:a[@href and starts-with(@href, "./")]',
+            namespaces={"x": "http://www.w3.org/1999/xhtml"},
+        ):
+            links.append(node.attrib["href"])
+
+        # verifies every 10th page url in each book
+        for link in links[::10]:
+
+            links_replaced = link.replace("./", f"{s3_base_url}{s3_archive_folder}").replace(
+                ".xhtml", ".json"
+            )
+
+            res = requests.get(links_replaced)
+
+            s3_pages = urllib.request.urlopen(links_replaced).read()
```

Comment:
Minor: I think you're requesting the same link twice here, no? Could you remove this line and make the next one `s3_pages_jdata = res.json()`?

Reply:
@rnathuji I will have a look at it together with other changes I started doing. just to clean up the fixtures and code. I merge this and I will come back to this after redirects tests.